finesse/hardware/tc4820.py

```diff
@@ -0,0 +1,137 @@
+"""This module provides an interface to TC4820 temperature controllers.
+
+TODO: I've used float throughout this module, but a decimal type may be more appropriate
+and prevent rounding errors.
+"""
+import logging
+from typing import Any
+
+from serial import Serial
+
+
+class TC4820:
+    """An interface for TC4820 temperature controllers."""
+
+    MAX_POWER = 511
+    """The maximum value for the power property."""
+
+    def __init__(
+        self, port: str, baudrate: int = 115200, *serial_args: Any, **serial_kwargs: Any
+    ) -> None:
+        """Create a new TC4820.
+
+        Args:
+            port: Serial port name
+            baudrate: Serial port baudrate
+            serial_args: Extra arguments to Serial constructor
+            serial_kwargs: Extra keyword arguments to Serial constructor
+        """
+        self.serial = Serial(port, baudrate, *serial_args, **serial_kwargs)
+
+    def read_int(self) -> int:
+        """Read a message from the TC4820 and decode the number as a signed integer."""
+        # TODO: Handle errors?
+        message = self.serial.read_until(b"^").decode("ascii")
+
+        if len(message) != 8 or message[0] != "*" or message[-1] != "^":
+            raise ValueError("Malformed message received: {message}")
+
+        if message == "*XXXX60^":
+            raise ValueError("Bad checksum sent")
+
+        if message[6:7] != TC4820.checksum(message[1:5]):
+            raise ValueError("Bad checksum received")
+
+        # Turn the hex string into raw bytes, then convert the raw bytes to a signed int
+        raw_bytes = bytes.fromhex(message[1:5])
+        return int.from_bytes(raw_bytes, byteorder="big", signed=True)
+
+    def read_float(self) -> float:
+        """Read a message from the TC4820 and decode the number as a float.
+
+        Float values are encoded as 10x their value then converted to an int.
```

Comment:
How weird...

Reply:
Fixed-point arithmetic is more common with embedded devices because floating-point is much more complicated so you either need a better CPU or a dedicated floating-point unit, which makes things more expensive. It's a bit odd that it's base 10 though.